### PR TITLE
update the error message when OOM happens

### PR DESCRIPTION
**Pull request by SmartKeyerror**
By default, Greenplum tracks virtual memory internally for resource group using a memory auditor referred to as vmtracker, those methods defined in `vmem_tracker.h`.If we enabled resource group, memory allocation will go through an additional step——allocate memory from vmtracker first.

To implement this, Greenplum defines `MemoryAllocationStatus` to represent the state of memory allocation:

```cpp
typedef enum MemoryAllocationStatus
{
	MemoryAllocation_Success,
	MemoryFailure_VmemExhausted,
	MemoryFailure_SystemMemoryExhausted,
	MemoryFailure_QueryMemoryExhausted,
	MemoryFailure_ResourceGroupMemoryExhausted
} MemoryAllocationStatus;
```

When OOM happens, Greenplum should output different logs according to the state of memory allocation. But for now, the error message is confusing and misleading.

So this pr is try to make it clear what things really going on to users when OOM happens.

### CODE REVIEW

File: src/backend/utils/mmgr/memprot.c [hunk 1]
```diff
@@ -351,7 +351,7 @@ static void gp_failed_to_alloc(MemoryAllocationStatus ec, int en, int sz)
 		/* Hit MOP limit */
 		ereport(ERROR, (errcode(ERRCODE_GP_MEMPROT_KILL),
 				errmsg("Out of memory"),
-				errdetail("VM Protect failed to allocate %d bytes, %d MB available",
+				errdetail("Vmem limit reached, failed to allocate %d bytes from tracker, VM Tracker %d MB available",
```
Comment (adam8157) [inline src/backend/utils/mmgr/memprot.c:354]: "Vmem limit reached, failed to allocate %d bytes from the tracker, which has %d MB available"

File: src/backend/utils/mmgr/memprot.c [hunk 2]
```diff
@@ -361,7 +361,7 @@ static void gp_failed_to_alloc(MemoryAllocationStatus ec, int en, int sz)
 		/* Hit MOP limit */
 		ereport(ERROR, (errcode(ERRCODE_GP_MEMPROT_KILL),
 				errmsg("Out of memory"),
-				errdetail("Per-query VM protect limit reached: current limit is %d kB, requested %d bytes, available %d MB",
+				errdetail("Per-query memory limit reached: current limit is %d kB, requested %d bytes, VM Tracker %d MB per query available",
```
Comment (adam8157) [inline src/backend/utils/mmgr/memprot.c:364]: "Per-query memory limit reached: the current limit is %d kB, requested %d bytes, has %d MB available for this query"

File: src/backend/utils/mmgr/memprot.c [hunk 3]
```diff
@@ -370,16 +370,14 @@ static void gp_failed_to_alloc(MemoryAllocationStatus ec, int en, int sz)
...
-				)
+				errdetail("System memory limit reached, failed to allocate %d bytes from system", sz)
 		));
 	}
 	else if (ec == MemoryFailure_ResourceGroupMemoryExhausted)
 	{
 		ereport(ERROR, (errcode(ERRCODE_GP_MEMPROT_KILL),
 				errmsg("Out of memory"),
-				errdetail("Resource group memory limit reached")));
+				errdetail("Resource group memory limit reached, failed to allocate %d bytes from tracker", sz)));
```
Comment (adam8157) [inline src/backend/utils/mmgr/memprot.c:380]: `from the group`, we might change the implementation in the future.

**Review by adam8157 — APPROVED** [review]
It's good to give more clear error messages.

Please confirm if there are no official utilities depending on those specific words, and don't backport to previous versions.

File: src/backend/utils/mmgr/memprot.c (hunk 3 above)
Comment (SmartKeyerror) [reply src/backend/utils/mmgr/memprot.c]: Got it, revert this message

**Review by Aegeaner — APPROVED** [review]

**Comment by SmartKeyerror** [comment]
My local full dev pipeline are all green, so merge it.

**Pull request merged**